flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/shuffle/TieredResultPartitionTest.java

```diff
@@ -201,6 +203,43 @@ void testMetricsUpdateForBroadcastOnlyResultPartition() throws Exception {
         }
     }
 
+    @Test
+    void testRequestBuffersAfterPoolSizeDecreased() throws IOException {
+        final int numBuffers = 15;
+        BufferPool bufferPool = globalPool.createBufferPool(1, numBuffers);
+        TieredResultPartition resultPartition =
+                createTieredStoreResultPartitionWithStorageManager(2, bufferPool, false);
+        ResultSubpartitionView subpartitionView =
+                resultPartition.createSubpartitionView(0, new NoOpBufferAvailablityListener());
+
+        // Emits some records to fill up 10 buffers, including all 8 of the memory tiers and 2 of
+        // the disk tiers.
+        for (int i = 0; i < 10; i++) {
+            resultPartition.emitRecord(ByteBuffer.allocate(NETWORK_BUFFER_SIZE), 0);
+        }
+
+        IOMetrics ioMetrics = taskIOMetricGroup.createSnapshot();
+        assertThat(ioMetrics.getResultPartitionBytes()).hasSize(1);
+        ResultPartitionBytes partitionBytes =
+                ioMetrics.getResultPartitionBytes().values().iterator().next();
+        assertThat(partitionBytes.getSubpartitionBytes()).contains((long) 10 * NETWORK_BUFFER_SIZE);
+
+        // Halve the size of the buffer pool, then emit another piece of data.
+        bufferPool.setNumBuffers(numBuffers / 2);
+        resultPartition.emitRecord(ByteBuffer.allocate(NETWORK_BUFFER_SIZE), 0);
+
+        ioMetrics = taskIOMetricGroup.createSnapshot();
+        partitionBytes = ioMetrics.getResultPartitionBytes().values().iterator().next();
+        assertThat(partitionBytes.getSubpartitionBytes()).contains((long) 11 * NETWORK_BUFFER_SIZE);
+
+        subpartitionView.releaseAllResources();
+
+        assertThat(bufferPool.bestEffortGetNumOfUsedBuffers())
+                .isEqualTo(2 * (resultPartition.getNumberOfSubpartitions() + 1) + 1);
+
+        resultPartition.release();
+    }
```

Comment:
The purpose of this case is not very clear to me. What are the expected behaviors that we are verifying after decreasing the pool size?

Reply:
The purpose is to test the behavior of requesting buffers when buffers are occupied by the memory tier and at that time the pool size is decreased. 
Indeed it is not easy to understand, I'll refactor it.